incubator/hnc/controllers/object_controller.go

```diff
@@ -102,63 +104,96 @@ func (r *ObjectReconciler) ReconcileNamespace(ctx context.Context, log logr.Logg
-	// is the source and there's nothing to do.
+// getSourceNS gets source namespace if it's set in "labelInheritedFrom" label.
+// It returns an empty string if this instance is the source
+func getSourceNS(inst *unstructured.Unstructured) string {
 	labels := inst.GetLabels()
 	if labels == nil {
 		// this cannot be a copy
-		return false, nil
+		return ""
 	}
 	inheritedFrom, _ := labels[labelInheritedFrom]
-	if inheritedFrom == "" {
-		// this isn't a copy
-		return false, nil
+	return inheritedFrom
+}
+
+// getSourceInst gets source instance of this one. It returns nil if this instance isn't a copy.
+func (r *ObjectReconciler) getSourceInst(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, *unstructured.Unstructured, error) {
+	srcNS := getSourceNS(inst)
+	if srcNS == "" {
+		// this cannot be a copy
+		return false, nil, nil
 	}
 
-	// Read the source. If we find it, and the namespace is still an ancestor, no problem.
-	srcNm := types.NamespacedName{Namespace: inheritedFrom, Name: inst.GetName()}
+	// Get the source instance from source namespace
+	srcNm := types.NamespacedName{Namespace: srcNS, Name: inst.GetName()}
 	src := &unstructured.Unstructured{}
 	src.SetGroupVersionKind(inst.GroupVersionKind())
 	err := r.Get(ctx, srcNm, src)
-	if err == nil {
-		// The object exists. Is it still in an ancestor namespace?
-		if yes, err := r.isAncestor(inst.GetNamespace(), inheritedFrom); err != nil {
-			// This can fail if the names are bad. For now, take no action (ie don't
-			// delete). TODO: revisit this.
-			log.Error(err, "deleteObsolete")
-			return false, err
-		} else if yes {
-			// We found the object in an ancestor namespace; no need to delete it.
-			return false, nil
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("The source object no longer exists", "inheritedFrom", srcNS)
+			return true, nil, err
+		} else {
+			log.Info("Couldn't read", "source", srcNm)
+			return false, nil, err
 		}
-		log.Info("Will delete because the source namespace is no longer an ancestor", "inheritedFrom", inheritedFrom)
-	} else if errors.IsNotFound(err) {
-		log.Info("Will delete because the source object no longer exists", "inheritedFrom", inheritedFrom)
-	} else {
-		log.Info("Couldn't read", "source", srcNm)
-		return false, err
 	}
 
-	// Delete this instance.
-	if err := r.Delete(ctx, inst); err != nil {
-		log.Error(err, "Couldn't delete after source was missing")
+	return false, src, nil
+}
+
+// ensureCorrectAncestry checks to see if the given object is correct. It returns true if
+// it's correct and ready to be propagated. The incorrect object will be deleted
+// if it's obsolete; or marked as modified if it's changed.
+func (r *ObjectReconciler) ensureCorrectAncestry(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, error) {
+	// Find what we expect the source to be.
+	missing, src, err := r.getSourceInst(ctx, log, inst)
+	if err != nil && !missing {
```

Comment:
This should usually just be `if err != nil`, especially if you make the other change I've suggested above.

Reply:
done